File: GildedRose/GildedRose/GildedRose.m
//
//  GildedRose.m
//  GildedRose
//
//  Created by Isil Demir on 1/22/14.
//  Copyright (c) 2014 Isil Demir. All rights reserved.
//

#import "GildedRose.h"

@implementation GildedRose

@synthesize items;

- (id) init {
    
    self = [super init];
    
    if (self) {
        self.items = @[[Item itemWithName:@"+5 Dexterity Vest" sellIn:10 andQuality:20],
                       [Item itemWithName:@"Aged Brie" sellIn:2 andQuality:0],
                       [Item itemWithName:@"Elixir of the Mongoose" sellIn:5 andQuality:7],
                       [Item itemWithName:@"Sulfuras, Hand of Ragnaros" sellIn:0 andQuality:80],
                       [Item itemWithName:@"Backstage passes to a TAFKAL80ETC concert" sellIn:15 andQuality:20],
                       [Item itemWithName:@"Conjured Mana Cake" sellIn:3 andQuality:6]];
    }
    return self;
}


- (void) updateQuality {
    
    NSDictionary *itemCatalogue = [[[Constants alloc] init] itemCatalogue];
    
    for (Item __strong *item in self.items) {
        
        NSString *itemType = [itemCatalogue objectForKey:[item name]];
        
        if ([itemType isEqualToString:@"REGULAR"]) {
            
            RegularItem *tmp = [[RegularItem alloc] initWithItem:item];
            item = [tmp updateStatus];
        }
        else if ([itemType isEqualToString:@"AGED"]) {
            
            AgedItem *tmp = [[AgedItem alloc] initWithItem:item];
            item = [tmp updateStatus];
        }
        else if ([itemType isEqualToString:@"LEGENDARY"]) {
            
            LegendaryItem *tmp = [[LegendaryItem alloc] initWithItem:item];
            item = [tmp updateStatus];
        }
        else if ([itemType isEqualToString:@"SPECIAL"]) {
            SpecialItem *tmp = [[SpecialItem alloc] initWithItem:item];
            item = [tmp updateStatus];
        }
        else if ([itemType isEqualToString:@"CONJURED"]) {
            ConjuredItem *tmp = [[ConjuredItem alloc] initWithItem:item];
            item = [tmp updateStatus];
        }
        
    }
    
}



- (void) updateQuality_old {
    
    for (int i=0; i < [self.items count]; i++) {
        if (![[self.items[i] name] isEqualToString:@"Aged Brie"] && ![[self.items[i] name] isEqualToString:@"Backstage passes to a TAFKAL80ETC concert"]) {
            if ([self.items[i] quality] > 0) {
                if (![[self.items[i] name] isEqualToString:@"Sulfuras, Hand of Ragnaros"]) {
                    [self.items[i] setQuality:[self.items[i] quality] - 1];
                }
            }
        }
        else {
            if ([self.items[i] quality] < 50) {
                [self.items[i] setQuality:[self.items[i] quality] + 1 ];
                if ([[self.items[i] name] isEqualToString:@"Backstage passes to a TAFKAL80ETC concert"]) {
                    if ([self.items[i] sellIn] < 11) {
                        if ([self.items[i] quality] < 50) {
                            [self.items[i] setQuality:[self.items[i] quality] + 1];
                        }
                    }
                    if ([self.items[i] sellIn] < 6) {
                        
                        if ([self.items[i] quality] < 50) {
                            [self.items[i] setQuality:[self.items[i] quality] + 1];
                        }
                    }
                }
            }
        }
        if (![[self.items[i] name] isEqualToString:@"Sulfuras, Hand of Ragnaros"]) {
            [self.items[i] setSellIn:[self.items[i] sellIn] - 1];
        }
        if ([self.items[i] sellIn] < 0) {
            if (![[self.items[i] name] isEqualToString:@"Aged Brie"]) {
                if (![[self.items[i] name] isEqualToString:@"Backstage passes to a TAFKAL80ETC concert"]) {
                    if ([self.items[i] quality] > 0) {
                        if (![[self.items[i] name] isEqualToString:@"Sulfuras, Hand of Ragnaros"]) {
                            [self.items[i] setQuality:[self.items[i] quality] - 1];
                        }
                    }
                }
                else {
                    [self.items[i] setQuality:[self.items[i] quality] - [self.items[i] quality]];
                }
            }
            else {
                if ([self.items[i] quality] < 50) {
                    [self.items[i] setQuality:[self.items[i] quality] + 1];
                }
            }
        }
    }
    
}


@end

File: GildedRose-Ruby/gilded_rose.rb

require './item.rb'

class GildedRose

@items = []

  def initialize
    @items = []
    @items << Item.new("+5 Dexterity Vest", 10, 20)
    @items << Item.new("Aged Brie", 2, 0)
    @items << Item.new("Elixir of the Mongoose", 5, 7)
    @items << Item.new("Sulfuras, Hand of Ragnaros", 0, 80)
    @items << Item.new("Backstage passes to a TAFKAL80ETC concert", 15, 20)
    @items << Item.new("Conjured Mana Cake", 3, 6)
  end

  def update_quality
    properties = { 
      "+5 Dexterity Vest"                         => { sell_in: -1, quality: lambda { |s| (s == 0) ? -2 : -1 } }, 
      "Aged Brie"                                 => { sell_in: -1, quality: lambda { |s| (s == 0) ? 2 : 1 } },
      "Elixir of the Mongoose"                    => { sell_in: -1, quality: lambda { |s| (s == 0) ? -2 : -1 } },
      "Sulfuras, Hand of Ragnaros"                => { sell_in:  0, quality: lambda { |s| 0 } }, 
      "Backstage passes to a TAFKAL80ETC concert" => { sell_in: -1, quality: lambda { |s| case s when 0; -50 when 1..5; 3 when 6..10; 2 else 1 end } }, 
      "Conjured Mana Cake"                        => { sell_in: -1, quality: lambda { |s| (s == 0) ? -4 : -2 } }
    }

    @items.each do |item|
      properties.each do |key, property|
        if item.name == key
          if item.quality >= 0 and item.quality <= 50
            item.quality += property[:quality].call(item.sell_in)
            item.quality = [0, item.quality, 50].sort[1] 
            item.sell_in += property[:sell_in]
          end
        end
      end
    end
  end 

end
